### PR TITLE
add ParamSpec to asyncio.to_thread

### PR DESCRIPTION
**Pull request by graingert**

### CODE REVIEW

File: stdlib/asyncio/threads.pyi [hunk 1]
```diff
@@ -1,7 +1,9 @@
 import sys
 from typing import Any, Callable, TypeVar
```
Comment (graingert) [inline stdlib/asyncio/threads.pyi:2]:
```suggestion
from typing import Callable, TypeVar
```

**Comment by graingert** [comment]
hmm stubtest doesn't like the ParamSpec _P here:

```

C:\hostedtoolcache\windows\Python\3.9.9\x64\python.exe -m mypy.stubtest --check-typeshed --custom-typeshed-dir . --allowlist tests\stubtest_allowlists\py3_common.txt --allowlist tests\stubtest_allowlists\py39.txt --allowlist tests\stubtest_allowlists\win32.txt --allowlist tests\stubtest_allowlists\win32-py39.txt --ignore-positional-only
error: not checking stubs due to mypy build errors:
stdlib\asyncio\threads.pyi:9: error: Name "_P.args" is not defined
stdlib\asyncio\threads.pyi:9: error: Name "_P.kwargs" is not defined
```

**Comment by JelleZijlstra** [comment]
I just merged a PR upgrading stubtest to 0.930, let's see if that helps.

File: stdlib/asyncio/threads.pyi [hunk 2]
```diff
@@ -1,7 +1,9 @@
...
+from typing import Callable, TypeVar
+from typing_extensions import ParamSpec
 
-_T = TypeVar("_T")
+_P = ParamSpec("_P")
+_R = TypeVar("_R")
 
 if sys.version_info >= (3, 9):
-    async def to_thread(__func: Callable[..., _T], *args: Any, **kwargs: Any) -> _T: ...
+    async def to_thread(func: Callable[_P, _R], /, *args: _P.args, **kwargs: _P.kwargs) -> _R: ...
```
Comment (JelleZijlstra) [inline stdlib/asyncio/threads.pyi:9]:
```suggestion
    async def to_thread(__func: Callable[_P, _R], /, *args: _P.args, **kwargs: _P.kwargs) -> _R: ...
```
according to stubtest it's positional-only

Comment (graingert) [reply stdlib/asyncio/threads.pyi]: Ah stubtest doesn't support positional only `/`?

Comment (AlexWaygood) [reply stdlib/asyncio/threads.pyi]: Several type-checkers don't support it at all :/ https://github.com/python/typeshed/issues/4972

File: stdlib/asyncio/threads.pyi [hunk 3]
```diff
@@ -1,7 +1,9 @@
...
+from typing import Callable, TypeVar
+from typing_extensions import ParamSpec
 
-_T = TypeVar("_T")
+_P = ParamSpec("_P")
+_R = TypeVar("_R")
 
 if sys.version_info >= (3, 9):
-    async def to_thread(__func: Callable[..., _T], *args: Any, **kwargs: Any) -> _T: ...
+    async def to_thread(__func: Callable[_P, _R], /, *args: _P.args, **kwargs: _P.kwargs) -> _R: ...
```
Comment (graingert) [inline stdlib/asyncio/threads.pyi:9]:
```suggestion
    async def to_thread(__func: Callable[_P, _R], *args: _P.args, **kwargs: _P.kwargs) -> _R: ...
```

**Review by Akuli — APPROVED** [review]
It might be good to remove `# type: ignore` comments from other uses of `ParamSpec`, as they no longer seem to be necessary. CONTRIBUTING.md should be updated too.

**Comment by graingert** [comment]
> It might be good to remove `# type: ignore` comments from other uses of `ParamSpec`, as they no longer seem to be necessary. CONTRIBUTING.md should be updated too.

Can CI check for unused ignores?

**Comment by srittau** [comment]
> Can CI check for unused ignores?

Unfortunately not, because different type checkers might have different opinions about that.

**Pull request merged**